include/rmm/mr/device/sam_headroom_resource_adaptor.hpp

```diff
@@ -0,0 +1,159 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <rmm/cuda_device.hpp>
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/system_memory_resource.hpp>
+#include <rmm/resource_ref.hpp>
+
+namespace rmm::mr {
+/**
+ * @addtogroup device_resource_adaptors
+ * @{
+ * @file
+ */
+/**
+ * @brief Resource that adapts system memory resource to allocate memory with a headroom.
+ *
+ * When GPU memory is over-subscribed, system allocated memory (SAM) will migrate to the GPU
+ * and cannot migrate back, thus causing other CUDA calls to fail with out-of-memory errors. To
```

Comment:
*question/nit*: This seems to suggest that SAM wil _only_ migrate to the GPU if gpu memory is over-subscribed. But I think that is not the correct reading?

My interpretation of what we're trying to achieve is this:

> System allocated memory (SAM) can be migrated to the GPU, but is never migrated back the host. If GPU memory is over-subscribed, this can cause other CUDA calls to fail with out-of-memory errors. To work around this problem, when using a system memory resource, we reserve some GPU memory as headroom for other CUDA calls, and only conditionally migrate if the allocation would not eat in to the headroom. Since doing this check on every allocation can be expensive, checking whether migration should occur is only enabled above the specified allocation threshold size.

Thoughts?

Reply:
We don't directly control the migration, we can only set the preferred location. Changed some wording but let me know if it makes more sense.